File: src/dataset/javascript/saddle-points_exercism-ThomasZumsteg/claude/ClaudeSonnet4_saddle-points_v4.js
var Matrix = function(matrix) {
	this.rows = matrix.split("\n").map(function(row) { 
		return row.replace(/^\s+|\s+$/g,'').split(' ')
			.map(function(e) { return parseInt(e, 10)});
	});

	const numRows = this.rows.length;
	const numCols = this.rows[0].length;
	
	this.columns = new Array(numCols);
	for (let col = 0; col < numCols; col++) {
		this.columns[col] = new Array(numRows);
		for (let row = 0; row < numRows; row++) {
			this.columns[col][row] = this.rows[row][col];
		}
	}
	
	this.saddlePoints = saddlePoints(this);
}

function saddlePoints(matrix) {
	const numRows = matrix.rows.length;
	const numCols = matrix.rows[0].length;
	
	const row_max = new Array(numRows);
	for (let r = 0; r < numRows; r++) {
		let max = matrix.rows[r][0];
		for (let c = 1; c < numCols; c++) {
			if (matrix.rows[r][c] > max) {
				max = matrix.rows[r][c];
			}
		}
		row_max[r] = max;
	}
	
	const column_min = new Array(numCols);
	for (let c = 0; c < numCols; c++) {
		let min = matrix.columns[c][0];
		for (let r = 1; r < numRows; r++) {
			if (matrix.columns[c][r] < min) {
				min = matrix.columns[c][r];
			}
		}
		column_min[c] = min;
	}
	
	const saddlePoints = [];
	for (let r = 0; r < numRows; r++) {
		for (let c = 0; c < numCols; c++) {
			const element = matrix.rows[r][c];
			if (element === row_max[r] && element === column_min[c]) {
				saddlePoints.push([r, c]);
			}
		}
	}
	
	return saddlePoints;
}

module.exports = Matrix;

File: src/dataset/python/saddle-points_exercism-ThomasZumsteg/claude/ClaudeSonnet4_saddle_points_v4.py
"""Finds saddle points in a rectangular 2d array"""

def saddle_points(matrix):
    """Finds points that are the largest in their row
    and smallest in their column
    """
    if not matrix or not matrix[0]:
        return set()
    
    saddles = set()
    num_cols = len(matrix[0])
    
    # Pre-compute all row maxima and column minima
    row_maxima = [max(row) for row in matrix]
    col_minima = [min(matrix[r][c] for r in range(len(matrix))) for c in range(num_cols)]
    
    # Find saddle points by comparing pre-computed values
    for r in range(len(matrix)):
        row_max = row_maxima[r]
        for c in range(num_cols):
            if row_max == col_minima[c]:
                saddles.add((r, c))
    
    return saddles